message/pool/pool.go

```diff
@@ -0,0 +1,55 @@
+package pool
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+type Pool struct {
+	// This field needs to be the first in the struct to ensure proper word alignment on 32-bit platforms.
+	// See: https://golang.org/pkg/sync/atomic/#pkg-note-BUG
+	currentMessagesInPool atomic.Int64
+	messagePool           sync.Pool
+	maxNumMessages        uint32
+	maxMessageBufferSize  uint16
+}
+
+func New(maxNumMessages uint32, maxMessageBufferSize uint16) *Pool {
+	return &Pool{
+		maxNumMessages:       maxNumMessages,
+		maxMessageBufferSize: maxMessageBufferSize,
+	}
+}
+
+// AcquireMessage returns an empty Message instance from Message pool.
+//
+// The returned Message instance may be passed to ReleaseMessage when it is
+// no longer needed. This allows Message recycling, reduces GC pressure
+// and usually improves performance.
+func (p *Pool) AcquireMessage(ctx context.Context) *Message {
+	v := p.messagePool.Get()
+	if v == nil {
+		return NewMessage(ctx)
+	}
+	r := v.(*Message)
+	p.currentMessagesInPool.Dec()
+	r.ctx = ctx
+	return r
+}
+
+// ReleaseMessage returns req acquired via AcquireMessage to Message pool.
+//
+// It is forbidden accessing req and/or its' members after returning
+// it to Message pool.
+func (p *Pool) ReleaseMessage(req *Message) {
+	v := p.currentMessagesInPool.Load()
```

Comment:
There's a race condition here if two threads use the same pool (say with a `pool` with `maxNumMessages` = 1):

Thread 1:
```
v := p.currentMessagesInPool.Load() // v == 0
// thread is stopped here and Thread 2 is scheduled
```

Thread 2:
```
v := p.currentMessagesInPool.Load() // v == 0
...
req is put into the message pool
```
Thread 2 now resumes and we have 2 messages in the pool. 

Good ol' compare and swap loop must be used here for correctness:
```go
for true {
  v := p.currentMessagesInPool.Load()
  if v >= int64(p.maxNumMessages) {
    return
  }
  next := v + 1
  if (p.currentMessagesInPool.CAS(v, next)) {
    break
  }
}
```

Reply:
fixed